### PR TITLE
Show / Screen Manager

### LINKED ISSUE #862: [SDL 0134] Show Manager - iOS
Proposal: [Show Manager - iOS](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0134-ios-show-manager.md)

> This proposal is to create a manager (akin to the file manager) that will give developers a high-level API to handle text (including metadata from [SDL-0073](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0073-Adding-Metadata-Types.md)), graphics, and soft buttons.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/386

### PR DESCRIPTION
**Pull request by joeljfischer**
Fixes #862 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
Unit tests will be added, tested against Manticore / Sync Gen 3 (17276_DEVTEST)

### Summary
This PR implements SDL-0134, "Show Manager" proposal. This proposal helps developers by making the `Show` RPC API much friendlier to use. This proposal / PR focuses on text, image, and soft button updates. Additional proposals / PRs could be made to support additional screen related APIs (e.g. media clock timer / subscription buttons). It does this by presenting simple properties which can be set individually or batched in a `beginUpdates` / `endUpdates` combo.

##### Enhancements
* Adds `SDLScreenManager` with properties and batch updating text, images, and soft buttons.

### Tasks Remaining:
- [x] Unit tests
- [x] Edge Cases

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLSoftButtonManager.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ */
+typedef void(^SDLSoftButtonUpdateCompletionHandler)(NSError *__nullable error);
+
+@interface SDLSoftButtonManager : NSObject
+
+/**
+ HAX: This is necessary due to a Ford Sync 3 bug that doesn't like Show requests without a main field being set (it will accept them, but with a GENERIC_ERROR, and 10-15 seconds late...)
+ */
+@property (copy, nonatomic, nullable) NSString *currentMainField1;
+
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLSoftButtonManager.h]: This HAX doesn't seem to work with _Manticore_ when all 4 of the text-fields are set. The`textAndGraphicManager` update gets sent first and all the text-fields get set correctly. Then the `softButtonManager` update gets sent next and overwrites the `mainField2` and `mainField3` with an empty string. The result is text flashes for a fraction of a second in `mainField2` and `mainField3`.

File: SmartDeviceLink/SDLSoftButtonManager.m [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    } else {
+        self.inProgressUpdate.softButtons = [self sdl_softButtonsForCurrentState];
+    }
+
+    __weak typeof(self) weakSelf = self;
+    [self.connectionManager sendConnectionRequest:self.inProgressUpdate withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
+        __strong typeof(weakSelf) strongSelf = weakSelf;
+
+        NSLog(@"SoftButton update request: %@, response: %@, error: %@", request, response, error);
+        strongSelf.inProgressUpdate = nil;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLSoftButtonManager.m]: Delete the `NSLog`

File: SmartDeviceLink/SDLSoftButtonManager.m [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    }];
+}
+
+#pragma mark - Images
+
+- (void)sdl_uploadArtworks:(NSArray<SDLArtwork *> *)artworks withCompletionHandler:(void (^)(NSError *error))handler {
+    [self.fileManager uploadArtworks:artworks completionHandler:^(NSArray<NSString *> * _Nonnull artworkNames, NSError * _Nullable error) {
+        handler(error);
+    }];
+}
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLSoftButtonManager.m]: `sdl_uploadArtworks:handler` isn't being used.

File: SmartDeviceLink/SDLSoftButtonManager.m [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        }
+    }
+
+    // Upload initial images, then other state images
+    if (initialStatesToBeUploaded.count > 0) {
+        [self.fileManager uploadArtworks:[initialStatesToBeUploaded copy] completionHandler:^(NSArray<NSString *> * _Nonnull artworkNames, NSError * _Nullable error) {
+            [self sdl_updateWithCompletionHandler:nil];
+        }];
+    }
+    if (otherStatesToBeUploaded.count > 0) {
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLSoftButtonManager.m]: If I upload invalid image data, the `SDLSoftButtonManager` tries to send a soft button with `type` of `TEXT` but with no text. This `show` RPC fails with an `INVALID_DATA` result code. Is it possible to just return an error instead of sending the invalid soft button?

File: SmartDeviceLink/SDLSoftButtonManager.h (hunk 1 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLSoftButtonManager.h]: I'm not sure what's causing this for you, but the SoftButtonManager does not overwrite anything but `mainField1`. It's `mainField2` and 3 are nil, which does not change the current text. I've tested it and it seems fine.

Comment (joeljfischer) [reply SmartDeviceLink/SDLSoftButtonManager.h]: I see the bug now, trying to track it down

File: SmartDeviceLink/SDLSoftButtonManager.m (hunk 4 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLSoftButtonManager.m]: We can't return an error because this is the setter. I have added error logs however. Is this sufficient? I'm not quite sure the path your code is taking, the soft button shouldn't send in `sdl_textButtonsForCurrentState` if there's no text on _any_ of the buttons, it will just wait for the images to upload.

Comment (NicoleYarroch) [reply SmartDeviceLink/SDLSoftButtonManager.m]: It looks like the error occurs the second time the request is sent.
## 1st Update
One text-field set, one soft button with an invalid image (no text set for the button). This request returns `SUCCESS`.
```
2018-03-15 13:57:45.841869-0400 SDL_Smoke_Tests[10979:1087712] Text & Graphic update request: Show (request)
{
alignment = CENTERED;
graphic = "{\n    imageType = DYNAMIC;\n    value = \"sdl_BlankArt\";\n}";
mainField1 = "button with broken image";
mainField2 = "";
mainField3 = "";
mainField4 = "";
metadataTags = "{\n    mainField1 =     (\n    );\n}";
secondaryGraphic = "{\n    imageType = DYNAMIC;\n    value = \"sdl_BlankArt\";\n}";
}, response: Show (response)
{
resultCode = SUCCESS;
success = 1;
}, error: (null)
2018-03-15 13:57:45.842027-0400 SDL_Smoke_Tests[10979:1087703] [Proxy] Message received: Show (response)
{
resultCode = SUCCESS;
success = 1;
}
```
## 2nd Update
Same text-field and button as the 1st update. The second request returns `INVALID_DATA`.
```
2018-03-15 13:58:56.184273-0400 SDL_Smoke_Tests[10979:1094326] SoftButton update request: Show (request)
{
mainField1 = "button with broken image";
softButtons =     (
"{\n    isHighlighted = 0;\n    softButtonID = 0;\n    systemAction = \"DEFAULT_ACTION\";\n    type = TEXT;\n}"
);
}, response: Show (response)
{
resultCode = "INVALID_DATA";
success = 0;
}, error: Error Domain=com.sdl.lifecyclemanager.error Code=-1 "INVALID_DATA" UserInfo={NSLocalizedRecoverySuggestion=Have you tried turning it off and on again?, NSLocalizedDescription=INVALID_DATA, NSLocalizedFailureReason=}
```

File: SmartDeviceLink/SDLSoftButtonManager.h (hunk 1 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLSoftButtonManager.h]: Generic HMI issue filed: https://github.com/smartdevicelink/generic_hmi/issues/62

File: SmartDeviceLink/SDLSoftButtonManager.m (hunk 4 above)
Comment (joeljfischer) [reply SmartDeviceLink/SDLSoftButtonManager.m]: The 1st update you have there is a Text and Graphic Manager update, the second a soft button update. The first has nothing to do with soft buttons.

Comment (NicoleYarroch) [reply SmartDeviceLink/SDLSoftButtonManager.m]: Yes, the first update is correct. Why is the button being sent on the second update? I am sending the exact same request twice.

Comment (joeljfischer) [reply SmartDeviceLink/SDLSoftButtonManager.m]: Fixed
